hnix-store-remote/src/System/Nix/Store/Remote.hs

```diff
@@ -4,29 +4,269 @@
 {-# LANGUAGE OverloadedStrings   #-}
 {-# LANGUAGE RankNTypes          #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds           #-}
 {-# LANGUAGE TypeApplications    #-}
-module System.Nix.Store.Remote (
-    runStore
-  , syncWithGC
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE FlexibleContexts     #-}
+module System.Nix.Store.Remote
+{-
+  (
+    addToStore
+  , addToStoreNar
+  , addTextToStore
+  , addSignatures
+  , addIndirectRoot
+  , addTempRoot
+  , buildPaths
+  , ensurePath
+  , findRoots
+  , isValidPathUncached
+  , queryValidPaths
+  , queryAllValidPaths
+  , querySubstitutablePaths
+  , queryPathInfoUncached
+  , queryReferrers
+  , queryValidDerivers
+  , queryDerivationOutputs
+  , queryDerivationOutputNames
+  , queryPathFromHashPart
+  , queryMissing
   , optimiseStore
+  , runStore
+  , syncWithGC
   , verifyStore
-  ) where
+  )
+-}
+  where
 
 import           Control.Monad
+import           Control.Monad.State
+import           Control.Monad.Except
+import           Control.Monad.IO.Class    (liftIO)
+import qualified Data.Binary               as B
+import qualified Data.Binary.Put           as B
+import           Data.Maybe
+import qualified Data.ByteString.Char8     as BSC
+import qualified Data.ByteString.Lazy      as BSL
+import qualified Data.Map.Strict           as M
+import           Data.Proxy                (Proxy(Proxy))
+import           Data.Text                 (Text)
+import qualified Data.Text.Lazy                 as T
+import qualified Data.Text.Lazy.Encoding        as T
+
+import qualified System.Nix.Build          as Build
+import qualified Nix.Derivation            as Drv
+
+--import qualified System.Nix.GC             as GC
+import           System.Nix.Hash           (Digest, ValidAlgo)
+import           System.Nix.StorePath
+import           System.Nix.Hash
+import           System.Nix.Nar            (localPackNar, putNar, narEffectsIO, Nar)
+import           System.Nix.Util
+import           System.Nix.ValidPath
 
 import           System.Nix.Store.Remote.Types
 import           System.Nix.Store.Remote.Protocol
 import           System.Nix.Store.Remote.Util
 
+import Data.Text.Encoding (encodeUtf8)
+
 type RepairFlag = Bool
 type CheckFlag = Bool
+type CheckSigsFlag = Bool
+type SubstituteFlag = Bool
 
-syncWithGC :: MonadStore ()
-syncWithGC = void $ simpleOp SyncWithGC
+addToStore
+  :: forall a. (ValidAlgo a, NamedAlgo a)
+  => StorePathName -- BSL.ByteString
+  -> FilePath
+  -> Bool
+  -> Proxy a
+  -> (StorePath -> Bool)
+  -> RepairFlag
+  -> MonadStore StorePath
+addToStore name pth recursive algoProxy pfilter repair = do
+
+  -- TODO: Is this lazy enough? We need `B.putLazyByteString bs` to stream `bs`
+  bs  :: BSL.ByteString <- liftIO $ B.runPut . putNar <$> localPackNar narEffectsIO pth
+
+  runOpArgs AddToStore $ do
+    putText $ unStorePathName name
+
+    putBool $ not $ algoName @a `elem` ["sha256"] && recursive
```

Comment:
Is there a reason to use `elem` instead of `/=` here ?

Reply:
Fixed